builder/replica_pipeline_test.cc

```diff
@@ -0,0 +1,374 @@
+#include <functional>
+#include <iostream>
+#include <map>
+#include <string>
+
+#include "builder/pipeline.h"
+#include "builder/replica_pipeline.h"
+
+namespace slinky {
+namespace internal {
+
+namespace {
+
+// clang-format off
+std::function<pipeline()> multiple_outputs_replica =
```

Comment:
In case it helps clarify, what I was imagining here is something like:

```
TEST(replica, multiple_outputs) {
  auto p =
// BEGIN define_replica_pipeline() output
[]() -> ::slinky::pipeline {
  node_context ctx;
  auto in = buffer_expr::make(ctx, "in", sizeof(uint32_t), 3);
  auto sum_x = buffer_expr::make(ctx, "sum_x", sizeof(uint32_t), 2);
  auto y = var(ctx, "y");
  auto z = var(ctx, "z");
  auto _1 = variable::make(in->sym());
  auto _replica_fn_2 = [=](const buffer<const void>& i0, const buffer<void>& o0, const buffer<void>& o1) -> index_t {
    const buffer<const void>* input_buffers[] = {&i0};
    const buffer<void>* output_buffers[] = {&o0, &o1};
    const func::input inputs[] = {{in, {{(buffer_min(_1, 0)), (buffer_max(_1, 0))}, {(buffer_min(_1, 1)), (buffer_max(_1, 1))}, point(z)}}};
    const std::vector<var> outputs[] = {{y, z}, {z}};
    return ::slinky::internal::replica_pipeline_handler(input_buffers, output_buffers, inputs, outputs);
  };
  auto sum_xy = buffer_expr::make(ctx, "sum_xy", sizeof(uint32_t), 1);
  auto _fn_0 = func::make(std::move(_replica_fn_2), {{in, {{(buffer_min(_1, 0)), (buffer_max(_1, 0))}, {(buffer_min(_1, 1)), (buffer_max(_1, 1))}, point(z)}}}, {{sum_x, {y, z}}, {sum_xy, {z}}});
  _fn_0.loops({{z, 1, loop_mode::serial}});
  auto p = build_pipeline(ctx, {}, {in}, {sum_x, sum_xy}, {});
  return p;
}
// END define_replica_pipeline() output
  const int H = 20;
  const int W = 10;
  const int D = 5;
  buffer<int, 3> in_buf({W, H, D});
  init_random(in_buf);
  buffer<int, 2> sum_x_buf({H, D});
  buffer<int, 1> sum_xy_buf({D});
  sum_x_buf.allocate();
  sum_xy_buf.allocate();
  const raw_buffer* inputs[] = {&in_buf};
  const raw_buffer* outputs[] = {&sum_x_buf, &sum_xy_buf};
  eval_context eval_ctx;
  ASSERT_EQ(0, p.evaluate(inputs, outputs, eval_ctx));
  // No check of the the contents of outputs.
}
```

Reply:
But: now we are back to replicating logic (ie buffer allocations) from `pipeline_test.cc` elsewhere, which was something you expressly wanted to avoid earlier. Please clarify.

Reply:
Not all duplication is equal. IMO duplication is fine when duplicating things that are simple and unlikely to change (e.g. buffers and extents). It's less fine when we're duplicating things that we know are going to change a lot (the entire pipeline description, which I expect to be either completely or partially rewritten at some point).

Also: see my comment above: I want to try to find a way to automate collecting this information too, not just the pipeline description. That is probably going to be an important part of reproducing some bugs. Automating that would eliminate this duplication too. But that can wait for another change.

If something isn't clear about what I'm asking, please let me know and we can discuss further before changing it.